File: 프로그래머스/2/42747. H－Index/H－Index.py
def solution(citations):
    answer = 0
    for h in range(10001):
        upper = 0
        rest = 0
        for citation in citations:
            if h <= citation:
                upper += 1
            elif h >= citation:
                rest += 1
        if upper >= h and rest <= h:
            answer = max(answer, h)
    return answer

File: 프로그래머스/2/42747. H－Index/H－Index.java
import java.util.Arrays;
class Solution {
    public int solution(int[] citations) {
        int answer = 0;
        //정렬 후 뒤에서부터 h편 이상 인용되었다면 h가 답
        Arrays.sort(citations);
        
        return binarySearch(citations);
    }
    
    
    private int binarySearch(int[] arr) {
        int start = 0;
        int end = arr.length;
        int answer = 0;
        while(start <= end) {
            int mid = (start + end) / 2;
            
            int cnt = 0;
            for (int num : arr) {
                if (num >= mid) {
                    cnt++;
                }
            }
            if(cnt < mid) {
                end = mid - 1;
            } else {
                answer = mid;
                start = mid + 1;
            }
        }
        return answer;
    }
}